analyzers/tests/SonarAnalyzer.UnitTest/Helpers/SymbolHelperTest.cs

```diff
@@ -398,5 +398,185 @@ public void GetClassification_Method(MethodKind methodKind, string expected)
+        [DataRow("DerivedClosedGeneric", "MyInheritedAttribute", "MyDerivedInheritedAttribute", "MyDerivedNotInheritedAttribute", "MyUnannotatedAttribute")]
+        [DataRow("Implement")]
+        public void GetAttributesWithInherited_TypeSymbol(string className, params string[] expectedAttributes)
+        {
+            var code = $$"""
+                using System;
+
+                [AttributeUsage(AttributeTargets.All, Inherited = true)]
+                public class MyInheritedAttribute : Attribute { }
+
+                [AttributeUsage(AttributeTargets.All, Inherited = false)]
+                public class MyNotInheritedAttribute : Attribute { }
+
+                public class MyDerivedInheritedAttribute: MyInheritedAttribute { }
+
+                public class MyDerivedNotInheritedAttribute: MyNotInheritedAttribute { }
+
+                public class MyUnannotatedAttribute : Attribute { }
+
+                [MyInheritedAttribute]
+                [MyDerivedInheritedAttribute]
+                [MyNotInherited]
+                [MyDerivedNotInherited]
+                [MyUnannotatedAttribute]
+                public class BaseClass<T1>
+                {
+                }
+
+                [MyInheritedAttribute]
+                [MyDerivedInheritedAttribute]
+                [MyNotInherited]
+                [MyDerivedNotInherited]
+                [MyUnannotatedAttribute]
+                public interface IInterface
+                {
+                }
+
+                public class DerivedOpenGeneric<T1>: BaseClass<T1>
+                {
+                }
+
+                public class DerivedClosedGeneric: BaseClass<int>
+                {
+                }
+
+                public class Implement: IInterface
+                {
+                }
+
+                public class Program
+                {
+                    public static void Main()
+                    {
+                        new {{className.Replace(@"`1", "<int>")}}();
+                    }
+                }
+                """;
+            var compiler = new SnippetCompiler(code);
+            var objectCreation = compiler.GetNodes<ObjectCreationExpressionSyntax>().Should().ContainSingle().Subject;
+            if (compiler.GetSymbol<IMethodSymbol>(objectCreation) is { MethodKind: MethodKind.Constructor, ReceiverType: { } receiver })
+            {
+                var actual = receiver.GetAttributesWithInherited().Select(x => x.AttributeClass.Name).ToList();
+                actual.Should().BeEquivalentTo(expectedAttributes);
+            }
+            else
+            {
+                Assert.Fail("Constructor could not be found.");
+            }
+            // GetAttributesWithInherited should behave like MemberInfo.GetCustomAttributes from runtime reflection:
+            var assembly = compiler.EmitAssembly();
+            var type = assembly.GetType(className, throwOnError: true);
```

Comment:
We don't need to work with the assembly
```suggestion
            var type = compiler.EmitAssembly().GetType(className, throwOnError: true);
```